src/MixedQuoter.sol

```diff
@@ -0,0 +1,311 @@
+            V3SmartRouterHelper.getStableInfo(factoryStable, params.tokenIn, params.tokenOut, params.flag);
+        amountOut = IStableSwap(swapContract).get_dy(i, j, params.amountIn);
+    }
+
+    /**
+     * Mixed *************************************************
+     */
+    function quoteMixedExactInput(
+        address[] calldata paths,
+        bytes calldata actions,
+        bytes[] calldata params,
+        uint256 amountIn
+    ) external override returns (uint256 amountOut) {
+        uint256 numActions = actions.length;
+        if (numActions == 0) revert NoActions();
+        if (numActions != params.length || numActions != paths.length - 1) revert InputLengthMismatch();
+
+        for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+            address tokenIn = paths[actionIndex];
+            address tokenOut = paths[actionIndex + 1];
+            if (tokenIn == tokenOut) revert InvalidPath();
+
+            uint256 action = uint256(uint8(actions[actionIndex]));
+            if (action == MixedQuoterActions.V2_EXACT_INPUT_SINGLE) {
+                // params[actionIndex] is zero bytes
+                amountIn = quoteExactInputSingleV2(
+                    QuoteExactInputSingleV2Params({tokenIn: tokenIn, tokenOut: tokenOut, amountIn: amountIn})
+                );
+            } else if (action == MixedQuoterActions.V3_EXACT_INPUT_SINGLE) {
+                // params[actionIndex]: abi.encode(fee)
+                uint24 fee = abi.decode(params[actionIndex], (uint24));
+                (uint256 _amountOut,,,) = quoteExactInputSingleV3(
+                    QuoteExactInputSingleV3Params({
+                        tokenIn: tokenIn,
+                        tokenOut: tokenOut,
+                        amountIn: amountIn,
+                        fee: fee,
+                        sqrtPriceLimitX96: 0
+                    })
+                );
+                amountIn = _amountOut;
+            } else if (action == MixedQuoterActions.V4_CL_EXACT_INPUT_SINGLE) {
+                QuoterMixedV4ExactInputSingleParams memory clParams =
+                    abi.decode(params[actionIndex], (QuoterMixedV4ExactInputSingleParams));
+                (tokenIn, tokenOut) = convertWETHToV4NativeCurency(clParams.poolKey, tokenIn, tokenOut);
+                bool zeroForOne = tokenIn < tokenOut;
+                checkV4PoolKeyCurrency(clParams.poolKey, zeroForOne, tokenIn, tokenOut);
+                (int128[] memory deltaAmounts,,) = clQuoter.quoteExactInputSingle(
+                    ICLQuoter.QuoteExactSingleParams({
+                        poolKey: clParams.poolKey,
+                        zeroForOne: zeroForOne,
+                        exactAmount: amountIn.toUint128(),
+                        sqrtPriceLimitX96: 0,
+                        hookData: clParams.hookData
+                    })
+                );
+                amountIn = deltaAmounts[1].toUint256();
```

Comment:
how do u make sure amountOut is always token1 ?

Reply:
i mean, should we do 

`zeroForOne ? amt1 : amt0`

Reply:
yes, correct